File: monte-carlo-random-walk-D.R

# One sample of a random walk
random_walk = function(nsteps) {
  xo = 0 
  moeda = c(-1,1)
  alldx = sample(moeda, size=nsteps, replace=TRUE)
  x_t   =  c(xo, cumsum(alldx) )
  return( x_t )
}

# Estimate the prob from the frequency
obtain_relative_frequency = function(v) {
 df  = data.frame( table(v) ) 
 vx  = as.integer( as.vector(df$v) ) # ~ unique()
 fa  = as.integer( as.vector(df$Freq) ) 
 vfr = fa/sum(fa)  #table(vx)/length(vx) 
 return( data.frame(vx,vfr) )
}

# To obtain D from Monte Carlo simulation of random walks
D_from_random_walk = function(nsteps,n_samp){

  many_x_t = replicate(n_samp,random_walk(nsteps) ) 
  v        = many_x_t[nsteps,]    
  output = obtain_relative_frequency( v )   
  x      = output$vx
  fr     = output$vfr
  x2 = sum( (x^2)*fr ) 
  D  = x2/(2*nsteps)  
  #print( data.frame(nsteps,n_samp,D) )
  return( D )
}

tmax  = 100
vns   = 10^(1:3)
nexperiments = 50

listres=list()
for( i in 1:nexperiments){
 vtmax = rep(tmax,length(vns))
 res   = mapply( function(x, y) D_from_random_walk(x,y), vtmax,vns)
 listres[[i]]=res
}
matres = as.matrix(as.data.frame(listres))  #list ---> matrix
                   
nameout=sprintf('random-walk%d-nexperiments%d-D-from-R.png',tmax,max(vns),nexperiments)
png( nameout )

matplot(matres,pch=16,xlab='',ylab='',xaxt='n')
atx    = axTicks(1) 
myatx  = seq(min(atx),max(atx),1)
labels = sapply( myatx, function(i) as.expression(bquote(10^ .(i)) ) )

axis(1,at=myatx,labels=labels, cex.axis=1.2 )
abline(h=0.5,lty=2,lwd=2)
mtext(side=1,cex=2,line=2.6,expression(n[samples]))
mtext(side=2,cex=2,line=2.4,'Diffusion coefficient D')
mtext(side=3,cex=2,line=0.1,'R')
legend('topright',legend=c('Analytical'),col='black',lty=2,lwd=2,pch=NA,bty='n')

dev.off()

File: monte-carlo-random-walk-D.py

import numpy as np
import collections as cl
from matplotlib import pylab as plt

# One sample of a random walk
def one_random_walk(nsteps=5):
  xo = 0
  moeda = [-1,1]
  alldx = np.cumsum(np.random.choice(moeda,size=nsteps))
  trajectory = np.concatenate( (xo,alldx), axis=None) 
  return trajectory
  
# Estimate the prob from the frequency
def obtain_relative_frequency(v):
 count = cl.Counter( v )
 x     = np.array( list( count.keys()   ) )
 fa    = np.array( list( count.values() ) )
 fr    = 1.0*fa/np.sum(fa)
 return x,fr

# To obtain D from Monte Carlo simulation of random walks
def D_from_random_walk(nsteps,ns):
  many_x_t  = [ one_random_walk(nsteps) for sample in range(ns) ]
  x_at_tmax = [ elem[-1] for elem in many_x_t ]
  
  vx, vfr = obtain_relative_frequency( x_at_tmax )  
  x2 = np.sum( (vx**2)*fr )
  D  = x2/(2*tmax) 
  
  return D


fig  = plt.figure()
ax   = plt.gca()

tmax = 100
nexperiments = 200
vns  = [ 10**j for j in range(1,4,1) ]

for i in range(nexperiments):  
  vD   = [ D_from_random_walk(tmax,nsamples) for nsamples in vns ]

  plt.plot(vns,vD,'o')
  ax.set_xscale('log')   

plt.title('Python')
plt.ylabel(r'Diffusion coefficient $D$', fontsize=14)
plt.xlabel('$n_{samples}$ ')
plt.axhline(y=0.5, color='k', linestyle='--',label='Analytical')
plt.legend()  

nameout=('random-walk%d-nexperiments%d-D-from-python.png')%(tmax,nexperiments)
plt.savefig( nameout )
plt.close()

